### PR TITLE
feat: create/remove frontmatter block [STU-174]

### PR DESCRIPTION
**Pull request by P0lip**

### CODE REVIEW

File: src/frontmatter/frontmatter.ts [hunk 1]
```diff
@@ -9,24 +9,41 @@ import { IFrontmatter, PropertyPath } from './types';
...
-  private readonly node: Unist.Literal | null;
-  private readonly root: Unist.Parent;
-  private readonly properties: T | null;
+  public readonly document: Unist.Parent;
+  private readonly node: Unist.Literal;
+  private properties: Partial<T> | null;
 
-  constructor(data: Unist.Parent | string) {
-    const root = typeof data === 'string' ? parseWithPointers(data).data : data;
+  constructor(data: Unist.Parent | string, dirty = true) {
```
Comment (P0lip) [inline src/frontmatter/frontmatter.ts]: I truly believe we should mutate the data by default. as this is what most tools that modify ASTs do.
Don't have any hard data to back up my statement, but the vast majority of them do alter the AST and this is totally fine as long as you keep the tree clean and don't leave it in an invalid state.
By leaving it in an invalid state I mean the following case:

- we modify given AST, but just a portion of it and we try to postpone all other needed operations
- these ops have to be manually flushed

In such case, until we execute these remaining ops, AST is left in stale state, which is usually not intended.

File: src/frontmatter/__tests__/fronmatter.spec.ts [hunk 2]
```diff
@@ -11,17 +11,24 @@ const FIXTURES_DIR = join(__dirname, './fixtures');
 const tags = fs.readFileSync(join(FIXTURES_DIR, 'tags.md'), 'utf-8');
 
 describe('Frontmatter', () => {
+  it('should clone given data if dirty is set to false', () => {
```
Comment (marbemac) [inline src/frontmatter/__tests__/fronmatter.spec.ts]: What is "dirty"?

Comment (P0lip) [reply src/frontmatter/__tests__/fronmatter.spec.ts]: The naming might be a bit unfortunate, but the intention of the `dirty` flag is to indicate whether you want to have your AST mutated or not.
'dirty' (default) = mutable
'clean' (dirty == false) = immutable

Comment (marbemac) [reply src/frontmatter/__tests__/fronmatter.spec.ts]: Maybe `mutate` would be more clear?

File: src/frontmatter/frontmatter.ts (hunk 1 above)
Comment (marbemac) [reply src/frontmatter/frontmatter.ts]: Not sure on wether this is correct or not, but our tree is immutable (all updates go through patch system), so having other libs like this one mutate it themselves doesn't work.

File: src/frontmatter/frontmatter.ts [hunk 3]
```diff
@@ -9,24 +9,41 @@ import { IFrontmatter, PropertyPath } from './types';
...
+    const root =
+      typeof data === 'string' ? parseWithPointers(data).data : dirty ? data : JSON.parse(JSON.stringify(data));
```
Comment (marbemac) [inline src/frontmatter/frontmatter.ts]: You could avoid deep cloning the entire tree by using something like immer?

Comment (P0lip) [reply src/frontmatter/frontmatter.ts]: I hesitated to use Immer as we touch almost every node.
Whenever a change is made, we update all positional info to keep lines in order to make sure diagnostics work well.
```
if (diff !== 0) {
  shiftLines(this.root, diff);
 }
```

Comment (P0lip) [reply src/frontmatter/frontmatter.ts]: Note to self: should add a test for new/removed block. I suppose we will need to take 2 extra lines into account and we don't now.

File: src/frontmatter/frontmatter.ts (hunk 1 above)
Comment (P0lip) [reply src/frontmatter/frontmatter.ts]: Yeah, in our case it's indeed immutable and given the fact this package is primarily used by Graphite, I believe we could make it immutable by default as opposed to the solution I proposed in the PR and still keep that flag if someone is willing to mutate the AST?

Comment (marbemac) [reply src/frontmatter/frontmatter.ts]: Perfect! Not mutating is also generally "safer".

File: src/frontmatter/frontmatter.ts (hunk 3 above)
Comment (marbemac) [reply src/frontmatter/frontmatter.ts]: Fair enough - this also isn't a hot path (updating frontmatter happens pretty infrequently) so performance isn't really an issue. 👍

Comment (P0lip) [reply src/frontmatter/frontmatter.ts]: Yeah, in general, I'm thinking whether we actually do need to shift these lines given the fact we actually send a patch and reparse the content anyway.
Shifting lines is potentially a nice addition, but we don't seem to need it at all and it adds complexity.
What's your take on that?
@marbemac

**Comment by P0lip** [comment]
@marbemac apart from lack of lines shifting for addition/removal case, I think this is good to go.

File: src/frontmatter/frontmatter.ts (hunk 3 above)
Comment (marbemac) [reply src/frontmatter/frontmatter.ts]: Agree! I think we can ditch it and simplify.

**Comment by marbemac** [comment]
Final step is yalc'ing this into studio and verifying it works. I just tried and markdown tags no longer work (both when frontmatter is not already there which obv didn't work before, but also adding new tags to markdown article that already has a tag doesn't work either).

When you have a moment, give it a go.

**Comment by marbemac** [comment]
Here is the relevant spot in Studio: https://github.com/stoplightio/studio/blob/release/may-2019/src/stores/tag.ts#L463. You can verify by navigating to a markdown article and changing the tags via the tag selector in top right of panel.

**Comment by P0lip** [comment]
@marbemac you need to update the code. Right now it's immutable, so you need to do it as follows:
```ts
 const fm = new Frontmatter(parsed);
```
and then in line 482
```ts
value: fm.document,
```

**Comment by marbemac** [comment]
Ahh, will try one sec.

**Review by marbemac — APPROVED** [review]

**Pull request merged**